File: lib/data/notification_classifier.dart
import 'models/notification_log.dart';

/// The verdict on one piece of notification text.
class ClassificationResult {
  const ClassificationResult({required this.status, this.reason, this.signals = const []});

  final NotificationStatus status;

  /// Plain-language explanation, null when nothing was flagged.
  final String? reason;

  /// The specific phrases that matched, so the detail screen can show the user
  /// what to look for next time rather than just asserting a verdict.
  final List<String> signals;
}

/// Deterministic keyword and regex matching over incoming notification text.
///
/// Runs entirely on the device with no network involved, which is what lets
/// the permission screen promise that notification text never leaves the phone.
///
/// TODO(ml): replace deterministic keyword check with TFLite classifier
/// Keep this class as the fallback when the model is unavailable, and keep the
/// [ClassificationResult.signals] contract: whatever decides, the user is still
/// owed a reason they can check.
abstract final class NotificationClassifier {
  /// Predatory-lending phrasing, paired with why it is a warning sign.
  ///
  /// Written in ordinary spelling. Both these keys and the incoming text are
  /// pushed through [_normalise] before comparison, so they cannot fall out of
  /// step with the squashing rule.
  static const _predatorySignals = <String, String>{
    'tanpa bi checking':
        'Menghindari pemeriksaan yang wajib pada pemberi pinjaman resmi.',
    'tanpa slip gaji': 'Melewati pengecekan kemampuan bayar.',
    'tanpa jaminan cair': 'Menjanjikan uang tanpa syarat yang masuk akal.',
    'tanpa survey': 'Melewati verifikasi yang biasa dilakukan lembaga resmi.',
    'langsung cair': 'Janji pencairan super cepat.',
    'cair 3 menit': 'Janji pencairan super cepat.',
    'cair 5 menit': 'Janji pencairan super cepat.',
    'cair hitungan menit': 'Janji pencairan super cepat.',
    'klik sekarang': 'Mendesak kamu memutuskan cepat.',
    'buruan': 'Mendesak kamu memutuskan cepat.',
    'limit kamu naik': 'Menawarkan utang yang tidak kamu minta.',
    'limit naik': 'Menawarkan utang yang tidak kamu minta.',
    'pinjaman disetujui': 'Mengaku menyetujui pinjaman yang tidak kamu ajukan.',
    'dana talangan': 'Istilah yang sering dipakai pinjaman ilegal.',
  };

  /// Due-date phrasing. These are recorded for awareness and never dismissed:
  /// a real bill is the user's own business.
  static final _billPatterns = <RegExp>[
    RegExp(r'jatuh tempo'),
    RegExp(r'bayar minimum'),
    RegExp(r'segera bayar'),
    RegExp(r'tagihan .*(jatuh|tempo|bayar)'),
  ];

  /// Lowercases and collapses runs of the same letter.
  ///
  /// The squashing defeats the common trick of padding words to slip past a
  /// filter, so "caiiiir" and "cair" compare equal. It is lossy on ordinary
  /// words too ("banned" becomes "baned"), which is harmless here because the
  /// result is only ever compared against text normalised the same way.
  ///
  /// Note this uses replaceAllMapped, not replaceAll. Dart's replaceAll takes
  /// the replacement literally, so the handoff's `r'$1'` form would insert the
  /// characters "$1" rather than the captured letter.
  static String _normalise(String text) {
    return text.toLowerCase().replaceAllMapped(
      RegExp(r'([a-z])\1+'),
      (match) => match[1]!,
    );
  }

  /// Classifies notification [text] from [sourceApp].
  static ClassificationResult classify(String text, {String? sourceApp}) {
    final normalised = _normalise(text);

    final matched = <String>[];
    final reasons = <String>[];
    for (final entry in _predatorySignals.entries) {
      if (normalised.contains(_normalise(entry.key))) {
        matched.add(entry.key);
        if (!reasons.contains(entry.value)) reasons.add(entry.value);
      }
    }

    if (matched.isNotEmpty) {
      return ClassificationResult(
        status: NotificationStatus.mencurigakan,
        reason: reasons.first,
        signals: matched,
      );
    }

    for (final pattern in _billPatterns) {
      if (pattern.hasMatch(normalised)) {
        return const ClassificationResult(
          status: NotificationStatus.tagihan,
          reason: 'Terdeteksi sebagai pengingat jatuh tempo.',
        );
      }
    }

    return const ClassificationResult(status: NotificationStatus.aman);
  }

  /// Builds a log entry from raw notification text.
  static NotificationLog toLog({
    required String id,
    required String sourceApp,
    required String text,
    required DateTime time,
  }) {
    final result = classify(text, sourceApp: sourceApp);
    return NotificationLog(
      id: id,
      time: time,
      sourceApp: sourceApp,
      snippet: NotificationLog.makeSnippet(text),
      status: result.status,
      reason: result.reason,
    );
  }
}

File: android/app/src/main/kotlin/com/pikir/pikir/NotificationClassifier.kt
package com.pikir.pikir

/** What the scanner concluded about one notification. */
enum class ScanStatus(val wire: String) {
    /** Matched the predatory-lending patterns. Dismissed and replaced. */
    SUSPICIOUS("mencurigakan"),

    /** A due date. Recorded for awareness; the notification is left alone. */
    BILL("tagihan"),

    /** Nothing matched. Left completely untouched. */
    SAFE("aman"),
}

data class ScanResult(
    val status: ScanStatus,
    val reason: String? = null,
    val signals: List<String> = emptyList(),
)

/**
 * Deterministic keyword and regex matching over notification text.
 *
 * Runs entirely on the device with no network involved, which is what lets the
 * permission screen promise that notification text never leaves the phone.
 *
 * This mirrors lib/data/notification_classifier.dart. The two are kept in step
 * by hand for now; when the model lands, both should call it rather than
 * drifting apart.
 *
 * TODO(ml): replace deterministic keyword check with TFLite classifier
 */
object NotificationClassifier {

    /** Predatory-lending phrasing, paired with why it is a warning sign. */
    private val predatorySignals = linkedMapOf(
        "tanpa bi checking" to
            "Menghindari pemeriksaan yang wajib pada pemberi pinjaman resmi.",
        "tanpa slip gaji" to "Melewati pengecekan kemampuan bayar.",
        "tanpa jaminan cair" to "Menjanjikan uang tanpa syarat yang masuk akal.",
        "tanpa survey" to "Melewati verifikasi yang biasa dilakukan lembaga resmi.",
        "langsung cair" to "Janji pencairan super cepat.",
        "cair 3 menit" to "Janji pencairan super cepat.",
        "cair 5 menit" to "Janji pencairan super cepat.",
        "cair hitungan menit" to "Janji pencairan super cepat.",
        "klik sekarang" to "Mendesak kamu memutuskan cepat.",
        "buruan" to "Mendesak kamu memutuskan cepat.",
        "limit kamu naik" to "Menawarkan utang yang tidak kamu minta.",
        "limit naik" to "Menawarkan utang yang tidak kamu minta.",
        "pinjaman disetujui" to "Mengaku menyetujui pinjaman yang tidak kamu ajukan.",
        "dana talangan" to "Istilah yang sering dipakai pinjaman ilegal.",
    )

    /**
     * Due-date phrasing. These are recorded for awareness and never dismissed:
     * a real bill is the user's own business.
     */
    private val billPatterns = listOf(
        Regex("jatuh tempo"),
        Regex("bayar minimum"),
        Regex("segera bayar"),
        Regex("tagihan .*(jatuh|tempo|bayar)"),
    )

    private val repeatedLetter = Regex("([a-z])\\1+")

    /**
     * Lowercases and collapses runs of the same letter.
     *
     * Defeats the common trick of padding words to slip past a filter, so
     * "caiiiir" and "cair" compare equal. Lossy on ordinary words too, which is
     * harmless because both sides of every comparison are normalised the same
     * way.
     */
    private fun normalise(text: String): String =
        repeatedLetter.replace(text.lowercase()) { it.groupValues[1] }

    fun classify(text: String): ScanResult {
        val normalised = normalise(text)

        val matched = mutableListOf<String>()
        val reasons = mutableListOf<String>()
        for ((phrase, reason) in predatorySignals) {
            if (normalised.contains(normalise(phrase))) {
                matched.add(phrase)
                if (!reasons.contains(reason)) reasons.add(reason)
            }
        }

        if (matched.isNotEmpty()) {
            return ScanResult(ScanStatus.SUSPICIOUS, reasons.first(), matched)
        }

        for (pattern in billPatterns) {
            if (pattern.containsMatchIn(normalised)) {
                return ScanResult(
                    ScanStatus.BILL,
                    "Terdeteksi sebagai pengingat jatuh tempo.",
                )
            }
        }

        return ScanResult(ScanStatus.SAFE)
    }

    /** Trims text the way the log stores it. Also used for log lines. */
    fun snippet(text: String, limit: Int = 120): String {
        val collapsed = text.replace(Regex("\\s+"), " ").trim()
        if (collapsed.length <= limit) return collapsed
        return collapsed.substring(0, limit - 1) + "…"
    }
}
